File: samples/src/main/java/com/ricoh/livestreaming/sample/base/JwtAccessToken.kt
/*
 * Copyright 2022 RICOH Company, Ltd. All rights reserved.
 */

package com.ricoh.livestreaming.sample.base

import android.util.Base64
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import io.jsonwebtoken.security.Keys
import java.nio.charset.StandardCharsets
import java.util.*

class JwtAccessToken {
    companion object {
        /**
         * サンプルのためクライアントサイドに生成関数を追加していますが、
         * AccessTokenはアプリバックエンドを用意して生成してください。
         */
        fun createAccessToken(
                clientSecret: String,
                roomId: String,
                roomSpec: RoomSpec): String {
            val key = Keys.hmacShaKeyFor(clientSecret.toByteArray(StandardCharsets.UTF_8))

            val connectionId = Base64.encodeToString(Math.random().toString().toByteArray(), Base64.DEFAULT)
                    .replace("=", "")
                    .replace("+", "")
                    .replace("/", "")
                    .replace("\n", "")
            val cal: Calendar = Calendar.getInstance()
            cal.add(Calendar.MINUTE, -30)
            val nbf = cal.timeInMillis / 1000
            cal.add(Calendar.HOUR, 1)
            val exp = cal.timeInMillis / 1000

            return Jwts.builder()
                    .claim("nbf", nbf)
                    .claim("exp", exp)
                    .claim("room_id", roomId)
                    .claim("room_spec", roomSpec.getSpec())
                    .claim("connection_id", connectionId)
                    .signWith(key, SignatureAlgorithm.HS256)
                    .compact()
        }
    }
}

File: unity-samples/unity-app-oculus/Assets/Scripts/JwtAccessToken.cs
using System;
using UnityEngine;

public static class JwtAccessToken
{
    /**
     * サンプルのためクライアントサイドに生成関数を追加していますが、
     * AccessTokenはアプリバックエンドを用意して生成してください。
     */
    public static string CreateAccessToken(
        string clientSecret,
        string roomId,
        RoomSpec roomSpec)
    {
        byte[] guid = Guid.NewGuid().ToByteArray();
        string connectionId = Convert.ToBase64String(guid, 0, guid.Length)
            .Replace("=", "")
            .Replace("+", "")
            .Replace("/", "");
        byte[] secretByteArray = System.Text.Encoding.UTF8.GetBytes(clientSecret);

        using (var keys = new AndroidJavaClass("io.jsonwebtoken.security.Keys"))
        using (var key = keys.CallStatic<AndroidJavaObject>("hmacShaKeyFor", secretByteArray))
        using (var algorithm = new AndroidJavaClass("io.jsonwebtoken.SignatureAlgorithm"))
        using (var hs256 = algorithm.GetStatic<AndroidJavaObject>("HS256"))
        using (var jwts = new AndroidJavaClass("io.jsonwebtoken.Jwts"))
        using (var builder = jwts.CallStatic<AndroidJavaObject>("builder"))
        {
            DateTimeOffset nbf = DateTimeOffset.Now.AddMinutes(-30);

            builder.Call<AndroidJavaObject>("claim", "nbf", Utils.GetJavaObject(nbf.ToUnixTimeSeconds()));
            builder.Call<AndroidJavaObject>("claim", "exp", Utils.GetJavaObject(nbf.AddHours(1).ToUnixTimeSeconds()));
            builder.Call<AndroidJavaObject>("claim", "room_id", Utils.GetJavaObject(roomId));
            builder.Call<AndroidJavaObject>("claim", "connection_id", Utils.GetJavaObject("AndroidUnityQuest" + connectionId));
            builder.Call<AndroidJavaObject>("claim", "room_spec", roomSpec.GetSpecHashMap());
            builder.Call<AndroidJavaObject>("signWith", key, hs256);
            return builder.Call<string>("compact");
        }
    }

}
